tests/system/test_bucket.py

```diff
@@ -0,0 +1,777 @@
+    )
+    buckets_to_delete.append(created)
+
+    with_user_project = storage_client.bucket(
+        new_bucket_name, user_project=user_project
+    )
+
+    # Exercise bucket ACL w/ userProject
+    acl = with_user_project.acl
+    acl.reload()
+    acl.all().grant_read()
+    acl.save()
+    assert "READER" in acl.all().get_roles()
+
+    del acl.entities["allUsers"]
+    acl.save()
+    assert not acl.has_entity("allUsers")
+
+    # Exercise default object ACL w/ userProject
+    doa = with_user_project.default_object_acl
+    doa.reload()
+    doa.all().grant_read()
+    doa.save()
+    assert "READER" in doa.all().get_roles()
+
+    # Exercise IAM w/ userProject
+    test_permissions = ["storage.buckets.get"]
+    found = with_user_project.test_iam_permissions(test_permissions)
+    assert found == test_permissions
+
+    policy = with_user_project.get_iam_policy()
+    viewers = policy.setdefault("roles/storage.objectViewer", set())
+    viewers.add(policy.all_users())
+    with_user_project.set_iam_policy(policy)
+
+
+def test_bucket_copy_blob(
+    storage_client, buckets_to_delete, blobs_to_delete, user_project,
+):
+    payload = b"DEADBEEF"
+    new_bucket_name = _helpers.unique_name("copy-blob")
+    created = _helpers.retry_429_503(storage_client.create_bucket)(new_bucket_name)
+    buckets_to_delete.append(created)
+    assert created.name == new_bucket_name
+
+    blob = created.blob("CloudLogo")
+    blob.upload_from_string(payload)
+    blobs_to_delete.append(blob)
+
+    new_blob = _helpers.retry_bad_copy(created.copy_blob)(
+        blob, created, "CloudLogoCopy"
+    )
+    blobs_to_delete.append(new_blob)
+
+    copied_contents = new_blob.download_as_bytes()
+    assert copied_contents == payload
+
+
+def test_bucket_copy_blob_w_user_project(
+    storage_client, buckets_to_delete, blobs_to_delete, user_project,
+):
+    payload = b"DEADBEEF"
+    new_bucket_name = _helpers.unique_name("copy-w-requester-pays")
+    created = _helpers.retry_429_503(storage_client.create_bucket)(
+        new_bucket_name, requester_pays=True
+    )
+    buckets_to_delete.append(created)
+    assert created.name == new_bucket_name
+    assert created.requester_pays
+
+    blob = created.blob("simple")
+    blob.upload_from_string(payload)
+    blobs_to_delete.append(blob)
+
+    with_user_project = storage_client.bucket(
+        new_bucket_name, user_project=user_project
+    )
+
+    new_blob = _helpers.retry_bad_copy(with_user_project.copy_blob)(
+        blob, with_user_project, "simple-copy"
+    )
+    blobs_to_delete.append(new_blob)
+
+    assert new_blob.download_as_bytes() == payload
+
+
+def test_bucket_copy_blob_w_generation_match(
+    storage_client, buckets_to_delete, blobs_to_delete,
+):
+    payload = b"DEADBEEF"
+    new_bucket_name = _helpers.unique_name("generation-match")
+    created = _helpers.retry_429_503(storage_client.create_bucket)(
+        new_bucket_name, requester_pays=True
```

Comment:
I think we can remove `requester_pays=True` from this test case since we're just testing copy blob with generation match. Curious why this isn't causing a bad request without `user_project` set?

Reply:
34466b6

Reply:
34466b6